### DISCUSSION TITLE
How to make a model works like JSON Schema's `patternProperties`?

### DISCUSSION THREAD

**Discussion by fy0** [pydantic/pydantic#2668] [Question]
```json
{
  "type": "object",
  "patternProperties": {
    "^S_": { "type": "string" },
    "^I_": { "type": "integer" }
  },
  "additionalProperties": false
}
```

It's a example from [Understanding JSON Schema](http://json-schema.org/understanding-json-schema/reference/object.html#pattern-properties)

This schema can validate that json:
```json
{ "S_25": "This is a string", "I_2": 456 }
```

I tried to write a equivalent model, but failed. It's my code:

```
class TestModel(BaseModel):
    __root__: Union[
        Dict[constr(regex="^S_"), str],
        Dict[constr(regex="^I_"), int],
    ]
```

**Comment by fy0** [comment]
Maybe like this :

```
class TestModel(BaseModel):
    __root__: DictWitKVPair[
        Union[
            KVPair[constr(regex="^S_"), str],
            KVPair[constr(regex="^I_"), int],
        ],
    ]
```

but `DictWitKVPair` type is not exists.

**Reply by PrettyWood** [reply]
Yep probably need to create a custom type for this!

**Reply by fy0** [reply]
> Yep probably need to create a custom type for this!

Hi, I wrote a demo:

https://gist.github.com/fy0/4ccd3d3234b0a8fb1b5baaea436232d2

The code used a monkey patch because Generic is incompatible with `BaseModel.schema()` and to read sub_fields.

I have no time to make a PR, welcome someone to do it.
